src/test/java/guide/UsingHttpsTest.java

```diff
@@ -56,29 +68,133 @@ public void shouldPassWhenExpectedStubDidHappen() throws GeneralSecurityExceptio
 
     @Test
     public void shouldBePossibleToSpecifyKeyStoresWithStandardProperties() throws GeneralSecurityException, IOException {
+        // use a different key store
+        System.setProperty("javax.net.ssl.keyStore", "build/resources/test/keystore_server_test");
+        System.setProperty("javax.net.ssl.keyStorePassword", "secret");
 
-        System.setProperty("javax.net.ssl.trustStore","build/resources/main/keystore_server");
-
-        System.setProperty("javax.net.ssl.keyStore","build/resources/main/truststore_server");
+        server.run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen();
 
-        System.setProperty("javax.net.ssl.trustStorePassword","secret");
+        KeyStore trustStore = createKeyStore("keystore_server_test_cert");
+        HttpResponse execute = sslReadyHttpClient(trustStore).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
 
-        System.setProperty("javax.net.ssl.keyStorePassword","secret");
+        assertThat(execute.getStatusLine().getStatusCode(), equalTo(200));
+        ensureHttp(server).gotStubsCommitmentsDone();
+    }
 
+    @Test
+    public void shouldFailWhenClientDoesNotConfigureTrust() {
         server.run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        try {
+            sslReadyHttpClient(null).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as client does not trust server's TLS certificate");
+        } catch (Exception e) {
+            assertThat(e, isA(SSLHandshakeException.class));
+        }
+
+        ensureHttp(server).gotStubsCommitmentsDone();
+    }
+
+    @Test
+    public void shouldFailWhenServerRequiresAuthentication() throws IOException {
+        server.clientAuth(Files.readAllBytes(Path.of("build/resources/test/keystore_server_test_cert")), "changeit").run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        try {
+            sslReadyHttpClient().execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as server requires client authentication");
+        } catch (Exception e) {
+            assertThat(e, isA(IOException.class));
```

Comment:
Is `org.apache.http.NoHttpResponseException` the right exception to be expected here?

Reply:
It is actually `SocketException` as it fails during socket connection. I changed to that instead.

Reply:
It maybe `SocketException` or one of `SSLException`, depending on the timing (not deterministic). That's why I chose to use `IOException`.